geo/src/algorithm/line_intersection.rs

```diff
@@ -0,0 +1,663 @@
+    let max_x1 = if q.start.x > q.end.x {
+        q.start.x
+    } else {
+        q.end.x
+    };
+    let max_y1 = if q.start.y > q.end.y {
+        q.start.y
+    } else {
+        q.end.y
+    };
+
+    let int_min_x = if min_x0 > min_x1 { min_x0 } else { min_x1 };
+    let int_max_x = if max_x0 < max_x1 { max_x0 } else { max_x1 };
+    let int_min_y = if min_y0 > min_y1 { min_y0 } else { min_y1 };
+    let int_max_y = if max_y0 < max_y1 { max_y0 } else { max_y1 };
+
+    let mid_x = (int_min_x + int_max_x) / (F::one() + F::one());
+    let mid_y = (int_min_y + int_max_y) / (F::one() + F::one());
+
+    // condition ordinate values by subtracting midpoint
+    let p1x = p.start.x - mid_x;
+    let p1y = p.start.y - mid_y;
+    let p2x = p.end.x - mid_x;
+    let p2y = p.end.y - mid_y;
+    let q1x = q.start.x - mid_x;
+    let q1y = q.start.y - mid_y;
+    let q2x = q.end.x - mid_x;
+    let q2y = q.end.y - mid_y;
+
+    // unrolled computation using homogeneous coordinates eqn
+    let px = p1y - p2y;
+    let py = p2x - p1x;
+    let pw = p1x * p2y - p2x * p1y;
+
+    let qx = q1y - q2y;
+    let qy = q2x - q1x;
+    let qw = q1x * q2y - q2x * q1y;
+
+    let x = py * qw - qy * pw;
+    let y = qx * pw - px * qw;
+    let w = px * qy - qx * py;
+
+    let x_int = x / w;
+    let y_int = y / w;
+
+    // check for parallel lines
+    if (x_int.is_nan() || x_int.is_infinite()) || (y_int.is_nan() || y_int.is_infinite()) {
+        None
+    } else {
+        // de-condition intersection point
+        Some(Coordinate {
+            x: x_int + mid_x,
+            y: y_int + mid_y,
+        })
+    }
+}
+
+/// This method computes the actual value of the intersection point.
+/// To obtain the maximum precision from the intersection calculation,
+/// the coordinates are normalized by subtracting the minimum
+/// ordinate values (in absolute value).  This has the effect of
+/// removing common significant digits from the calculation to
+/// maintain more bits of precision.
+fn proper_intersection<F: GeoFloat>(p: Line<F>, q: Line<F>) -> Coordinate<F> {
+    // Computes a segment intersection using homogeneous coordinates.
+    // Round-off error can cause the raw computation to fail,
+    // (usually due to the segments being approximately parallel).
+    // If this happens, a reasonable approximation is computed instead.
+    let mut int_pt = raw_line_intersection(p, q).unwrap_or_else(|| nearest_endpoint(p, q));
+
+    // NOTE: At this point, JTS does a `Envelope::contains(coord)` check, but confusingly,
+    // Envelope::contains(coord) in JTS is actually an *intersection* check, not a true SFS
+    // `contains`, because it includes the boundary of the rect.
+    if !(p.bounding_rect().intersects(&int_pt) && q.bounding_rect().intersects(&int_pt)) {
+        // compute a safer result
+        // copy the coordinate, since it may be rounded later
+        int_pt = nearest_endpoint(p, q);
+    }
+    return int_pt;
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
As well as these tests, my main confidence for correctness comes transitively from using this code and passing the [JTS suite of "Relate" tests](https://github.com/georust/jts-test-runner/tree/mkirk/relate).